### PR TITLE
First pass at saving, in particular Save-As emacs-like dialogue.

### PR DESCRIPTION
**Pull request by slevination**
@alexhsamuel 

First pass at using the cmd box that you "created".   Obviously, saving itself needs some work (atomicity, warning if overwriting an existing file, types)

### CODE REVIEW

**Comment by slevination** [comment]
@alexhsamuel 

please ignore for now.  Debugging and trying to merge with your branch.

**Comment by slevination** [comment]
Debugged textbox input stuff.  seems a little better now (at least no extra space at the end).

File: tbl/curses_keyboard.py [hunk 1]
```diff
@@ -41,12 +41,15 @@
...
-    127:    "BACKSPACE",
+    # 127:    "BACKSPACE",
     258:    "DOWN",
     259:    "UP",
     260:    "LEFT",
     261:    "RIGHT",
     262:    "HOME",
+    # SL:  At least in my setup, BACKSPACE=263 (0407)
+    # see: https://stackoverflow.com/questions/14103341/weird-key-values-printed-by-ncurses
+    curses.KEY_BACKSPACE: 'BACKSPACE',
```
Comment (alexhsamuel) [inline tbl/curses_keyboard.py:52]: Seems to be OS/X produces 127 while Linux produces `cuses.KEY_BACKSPACE` = 263. `curses.erasechar()`, which you use below, seems to return the correct one; let's use that.

File: tbl/curses_keyboard.py [hunk 2]
```diff
@@ -73,19 +76,29 @@
...
-                return "M-ESC"
+                return "M-ESC", None
             else:
                 meta = True
                 continue
         elif c == curses.ERR:
             # Not sure why we get these.
             continue
+        elif c == curses.erasechar():
+            return "BACKSPACE", None
```
Comment (alexhsamuel) [inline tbl/curses_keyboard.py:101]: I think this can be handled by `KEYS` above, right?

File: tbl/curses_keyboard.py [hunk 3]
```diff
@@ -73,19 +76,29 @@
...
         c = stdscr.getch()
+
         if c == ESC:
+            """
+            Note: the below code will take a while to return just ESC
+            if nothing after it is pressed.  This has to do with the way getch()
+            handles Esc-<sym> stuff.
+            """
+            if not process_escape_meta:
+                return "ESC", None
```
Comment (alexhsamuel) [inline tbl/curses_keyboard.py:91]: What's the use case for this?  With this enabled, meta keys and some other special keys will produce an immediate escape, likely followed later by some spurious character.

File: tbl/model.py [hunk 4]
```diff
@@ -17,13 +19,14 @@ def __init__(self, arr, name=None):
...
 
-    def __init__(self):
+    def __init__(self, filename):
         # Columns, in order.
         self.__cols         = []
         # Number of rows in the table, or None if no columns so far.
         self.__num_rows     = None
 
         self.__undo_info = []
+        self.filename = filename
```
Comment (alexhsamuel) [inline tbl/model.py:29]: Use a private attribute name?

File: tbl/screen.py [hunk 5]
```diff
@@ -196,24 +225,39 @@ def cmd_input(screen, stdscr, prompt=""):
...
-    }.get(c, c)
+
+    def _validate(c):
+        translate = {127: 8, # BACKSPACE -> C-h
+                     10: 7, # ENTER --> C-g
+                     343: 7 # RETURN --> C-g
+                     }
+
+        # Escape/C-g ==> abort, ala Emacs.
+        if c == 27 or c == 7:
```
Comment (alexhsamuel) [inline tbl/screen.py:236]: Err, I don't think emacs lets you abort the minibuffer by pressing escape.  This logic will prevent us from using meta combos while editing commands, and also will probably leave a spurious extra character available afters.  I suggest we only use C-g for this.

File: tbl/screen.py [hunk 6]
```diff
@@ -196,24 +225,39 @@ def cmd_input(screen, stdscr, prompt=""):
...
     box = curses.textpad.Textbox(win)
+    # box.stripspaces = True  # as far as I can tell, this does nothing and is on by default anyway.
     curses.curs_set(True)
     # Run it.
     try:
-        input = box.edit(translate)
+        input = box.edit(_validate).strip()
+        result = True, input
+    except EscapeInterrupt:
+        result = False, None
```
Comment (alexhsamuel) [inline tbl/screen.py:255]: Why not just return `None` for the input, or better yet, raise an exception?

**Pull request merged**